### PR TITLE
Implement base logic for `Hay` degradations

### PR DESCRIPTION
**Pull request by Isgotkowitz**
This PR implements the base logic for calculating gaseous dry matter loss in hayed crops.

### Context

Issue(s) closed by this pull request: addresses #1377

- [x] The [changelog](https://docs.google.com/spreadsheets/d/1U-3igxdstHTFb6k5dVcE5-x9IL0r6q18eUiwwytKZyA/edit#gid=0) has been updated.
      
How big are the changes in the PR? Would you nominate it for a major version upgrade?
- [ ] Yes
- [x] No

### What

This PR adds helper routines to calculate the amount of dry matter lost from all hayed crops. This includes routines to calculate the dry matter loss in the first 30 days of storage, and the loss from all days after thirty days of storage. It also implements `calculate_gaseous_dry_matter_loss` so that it has the same interface and behavior in `Hay` as it does in `Storage`. This will allow easier reuse of the `process_degradations` implementation that is in `Storage`.

It also adds some logic to `HarvestedCrop` which calculates and saves certain factors that are needed for calculating dry matter loss in `Hay`.

I have also updated the Feed Storage Scientific Documentation to have the updated logic for processing the dry matter loss in Hay.

### Why

These calculations are necessary to calculate the amount of dry matter being lost from `Hay` and all its children.

### How

The interface for `calculate_gaseous_dry_matter_loss` in `Storage` now includes the current time - which isn't needed there, but is for `Hay`'s implementation. By keeping these interfaces the same, it will be possible `Hay`'s implementation to cleanly override `Storage`.

### Test plan

Unit tests have been written or updated for all methods changed. I have looked through the test cases, and checked that all the numbers in the unit test cases give reasonable outputs for reasonable inputs.

### CODE REVIEW

File: tests/test_feed_storage_module/test_hay.py [hunk 1]
```diff
@@ -16,9 +20,88 @@ def hay() -> Hay:
...
+
+@pytest.mark.parametrize(
+    "days,expected",
+    [
+        (1, 950.665050408),
+        (10, 950.670114831),
+        (20, 950.676225724),
+        (30, 950.682926336),
+        (40, 950.682926336),
+        (100, 950.682926336),
```
Comment (Isgotkowitz) [inline tests/test_feed_storage_module/test_hay.py]: @KevinP-B I think something is off about these numbers. The "days" is the number of days the crop has been stored for, and the "expected is the cumulative amount of dry matter loss that has happened from this crop. It seems that too much loss is happening right at the beginning, with too little loss happening after that.

For context, this is happening in a crop that is 20% dry matter at time of storage and has 950 MJ / kg of total sensible heat generated.

**Review by KevinP-B — APPROVED** [review]
Code looks to effectively replicate scientific documentation and will make more functions interoperable between storage types.

File: RUFAS/routines/feed_storage/hay.py [hunk 2]
```diff
@@ -28,38 +39,111 @@ def __init__(self, capacity: float = float("inf")):
...
+        days_in_window = min(days_stored, 30)
+        dry_fraction = crop.initial_dry_matter_percentage * GeneralConstants.PERCENTAGE_TO_FRACTION
+        moisture_fraction = 1 - dry_fraction
+        numerator = moisture_fraction - FINAL_MOISTURE_FRACTION * dry_fraction * (1 - 0.004 * days_in_window)
+        denominator = dry_fraction * (14206 - 2433 * (0.004 * days_in_window) / (1 - 0.004 * days_in_window))
+        return crop.total_sensible_heat_generated + 2433 * numerator / denominator
+
+    def _calculate_subsequent_dry_matter_loss_to_gas(self, crop: HarvestedCrop, time: Time) -> float:
+        """
+        Calculates the amount of gaseous dry matter lost in a hayed crop in its first 30 days of storage.
```
Comment (KevinP-B) [inline RUFAS/routines/feed_storage/hay.py]: Should this say "after its first 30 days..."?

Comment (Isgotkowitz) [reply RUFAS/routines/feed_storage/hay.py]: Yes nice catch

File: docs/feed_storage.rst [hunk 3]
```diff
@@ -70,9 +70,9 @@ that occurred after the first 30 days of storage (kg).
 .. math::
 
 
-   L_{first 30 days} = Q + 2433(\frac{moisture - MF(1 - moisture)(\frac{1}{1 - 0.004 \cdot d})}{(1-moisture)(14206-2433[0.004 \cdot d])(\frac{1}{1-0.004 \cdot d})}) \tag{1.2.3}
+   L_{first 30 days} = \frac{Q + 2433 \cdot [moisture - \frac{MF(1 - moisture)}{1 - MF}]}{(1 - moisture)(14206 - \frac{2433 \cdot MF}{1 - MF})} \cdot \frac{min(30, d)}{30} \tag{1.2.3}
```
Comment (Isgotkowitz) [inline docs/feed_storage.rst:73]: This is my interpretation of this equation from IFSM:
image
I think it is correct, but the IFSM equation has an extra right parentheses in the denominator so there is still a little ambiguity.

Comment (KevinP-B) [reply docs/feed_storage.rst:73]: The IFSM equation is missing the left parenthesis for the term that should be (1-Mf). 

I'm not familiar enough with the syntax of math in python to know by looking, but so long as the /frac in curly brackets are nested as above with the parenthesis clarification AND adjacent parenthesis, i.e. ")(" are implicitly mutliplied, it looks good!

**Review by KevinP-B — APPROVED** [review]
These changes look good!

File: RUFAS/routines/feed_storage/harvested_crop.py [hunk 4]
```diff
@@ -146,3 +165,40 @@ def _estimate_maximum_effluent(self) -> float:
...
+
+    def _calculate_bale_density(self) -> float:
+        """
+        Calculates the bale density of hayed crop.
+
+        Returns
+        -------
+        float
+            Bale density of this crop in kg / cubic meter.
+
```
Comment (matthew7838) [inline RUFAS/routines/feed_storage/harvested_crop.py]: More unified to follow the format of having units in parenthesis.

File: RUFAS/routines/feed_storage/hay.py [hunk 5]
```diff
@@ -28,38 +39,115 @@ def __init__(self, capacity: float = float("inf")):
...
+    def bale_size(self) -> float:
         """
-        Calculate and return the density of the hay bale.
+        Return the size (diameter) of the hay bale.
 
         Returns
         -------
         float
-            The density of the hay bale.
+            The diameter of the hay bale in meters.
```
Comment (matthew7838) [inline RUFAS/routines/feed_storage/hay.py]: same comments as above

File: RUFAS/routines/feed_storage/hay.py [hunk 6]
```diff
@@ -28,38 +39,115 @@ def __init__(self, capacity: float = float("inf")):
...
+        time : Time
+            Time instance containing the time that loss should be processed up to.
 
         Returns
         -------
         float
-            The diameter of the hay bale in meters.
-        """
-        pass
+            Gaseous dry matter loss from the hayed crop that occurred in the first 30 days of storage in kg.
```
Comment (matthew7838) [inline RUFAS/routines/feed_storage/hay.py]: Same.
```suggestion
            Gaseous dry matter loss from the hayed crop that occurred in the first 30 days of storage (kg).
```

File: RUFAS/routines/feed_storage/hay.py [hunk 7]
```diff
@@ -28,38 +39,115 @@ def __init__(self, capacity: float = float("inf")):
...
+        days_in_window = min(days_stored, 30)
+        fraction_of_total_loss = days_in_window / 30
+        dry_fraction = crop.initial_dry_matter_percentage * GeneralConstants.PERCENTAGE_TO_FRACTION
+        moisture_fraction = 1 - dry_fraction
+        numerator = crop.total_sensible_heat_generated + 2433 * (
+            moisture_fraction - (FINAL_MOISTURE_FRACTION * dry_fraction) / (1 - FINAL_MOISTURE_FRACTION)
+        )
+        denominator = dry_fraction * (14206 - 2433 * FINAL_MOISTURE_FRACTION / (1 - FINAL_MOISTURE_FRACTION))
+        fraction_of_initial_dry_matter_lost = numerator / denominator * fraction_of_total_loss
+        return crop.initial_dry_matter_mass * fraction_of_initial_dry_matter_lost
```
Comment (matthew7838) [inline RUFAS/routines/feed_storage/hay.py]: A little spacing might improve readability.

File: RUFAS/routines/feed_storage/storage.py [hunk 8]
```diff
@@ -316,6 +318,9 @@ def calculate_dry_matter_loss_to_gas(
         no dry matter loss occurs. Alfalfa uses different parameters and limits for calculating dry matter loss,
         but the structure of the loss equation remains the same.
 
+        Note that the current time is not needed for calculating the dry matter loss to fermentation, but it allows the
+        interface to remain uniform across all implementations of `calculate_dry_matter_loss_to_gas`.
+
```
Comment (matthew7838) [inline RUFAS/routines/feed_storage/storage.py:323]: Can have a separate note section.

File: tests/test_feed_storage_module/test_hay.py [hunk 9]
```diff
@@ -16,9 +20,92 @@ def hay() -> Hay:
...
     assert hay.acceptable_crops == [
         CropCategory.ALFALFA,
         CropCategory.GRASS,
         CropCategory.SMALL_GRAIN,
     ]
+
+
+@pytest.mark.parametrize("stored_day,current_day,expect_loss", [(1, 1, False), (1, 10, True)])
+def test_calculate_dry_matter_loss_to_gas(
+    hay: Hay, harvested_crop: HarvestedCrop, mocker: MockerFixture, stored_day: int, current_day: int, expect_loss: bool
```
Comment (matthew7838) [inline tests/test_feed_storage_module/test_hay.py:49]: Missing docstrings.

**Review by matthew7838 — APPROVED** [review]
Looks good to me, some comments on docstrings and I think it's good to merge.
